### ISSUE TITLE
internal error: generated ill-typed AIR code

### ISSUE THREAD

**Issue by jonhnet** [verus-lang/verus#583]
Repro:
```
git clone git@github.com:vmware-labs/verified-betrfs.git
git checkout ae19f303cf1fb3d0cfd191f553da01b2764cdd07
cd verified-betrfs/Splinter
$verus --time --expand-errors --multiple-errors 5 src/bundle.rs --verify-module journal::LinkedJournal_v --verify-function impl&%1::tight_sub_disk
```

Observed:
```
note: verifying module journal::LinkedJournal_v

error: assertion failed
   --> src/journal/LinkedJournal_v.rs:526:29
    |
526 |                     assert( other.is_sub_disk(self) );
    |                             ^^^^^^^^^^^^^^^^^^^^^^^ assertion failed

error: assertion failed
   --> src/journal/LinkedJournal_v.rs:531:33
    |
531 | ...t( self.build_tight(self.next(root)).entries.contains_key(pptr.unwrap())...
    |       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ assertion failed

thread 'rustc' panicked at 'internal error: generated ill-typed AIR code: error 'error 'error 'error 'in call to vstd!set.impl&%0.contains.?, argument #4 has type "bundle!disk.GenericDisk_v.Address." when it should have type "Poly"' in expression '(vstd!set.impl&%0.contains.? TYPE%bundle!disk.GenericDisk_v.Address. TYPE%bundle!disk.GenericDisk_v.Address.
 (vstd!map.impl&%0.dom.? TYPE%bundle!disk.GenericDisk_v.Address. TYPE%bundle!disk.GenericDisk_v.Address.
  TYPE%bundle!journal.LinkedJournal_v.JournalRecord. TYPE%bundle!journal.LinkedJournal_v.JournalRecord.
  (Poly%vstd!map.Map<bundle!disk.GenericDisk_v.Address./bundle!journal.LinkedJournal_v.JournalRecord.>.
   (bundle!journal.LinkedJournal_v.DiskView./DiskView/entries (%Poly%bundle!journal.LinkedJournal_v.DiskView.
     other~152@
  )))
 ) k~12$
)'' in expression '(=>
 (= (bundle!journal.LinkedJournal_v.DiskView./DiskView/boundary_lsn (%Poly%bundle!journal.LinkedJournal_v.DiskView.
    (Poly%bundle!journal.LinkedJournal_v.DiskView. self~2@)
   )
  ) (bundle!journal.LinkedJournal_v.DiskView./DiskView/boundary_lsn (%Poly%bundle!journal.LinkedJournal_v.DiskView.
    other~152@
 )))
 (forall ((k~12$ bundle!disk.GenericDisk_v.Address.)) (!
   (=>
    (has_type (Poly%bundle!disk.GenericDisk_v.Address. k~12$) TYPE%bundle!disk.GenericDisk_v.Address.)
    (=>
     (vstd!set.impl&%0.contains.? TYPE%bundle!disk.GenericDisk_v.Address. TYPE%bundle!disk.GenericDisk_v.Address.
      (vstd!map.impl&%0.dom.? TYPE%bundle!disk.GenericDisk_v.Address. TYPE%bundle!disk.GenericDisk_v.Address.
       TYPE%bundle!journal.LinkedJournal_v.JournalRecord. TYPE%bundle!journal.LinkedJournal_v.JournalRecord.
       (Poly%vstd!map.Map<bundle!disk.GenericDisk_v.Address./bundle!journal.LinkedJournal_v.JournalRecord.>.
        (bundle!journal.LinkedJournal_v.DiskView./DiskView/entries (%Poly%bundle!journal.LinkedJournal_v.DiskView.
          other~152@
       )))
      ) k~12$
     )
     (vstd!set.impl&%0.contains.? TYPE%bundle!disk.GenericDisk_v.Address. TYPE%bundle!disk.GenericDisk_v.Address.
      (vstd!map.impl&%0.dom.? TYPE%bundle!disk.GenericDisk_v.Address. TYPE%bundle!disk.GenericDisk_v.Address.
       TYPE%bundle!journal.LinkedJournal_v.JournalRecord. TYPE%bundle!journal.LinkedJournal_v.JournalRecord.
       (Poly%vstd!map.Map<bundle!disk.GenericDisk_v.Address./bundle!journal.LinkedJournal_v.JournalRecord.>.
        (bundle!journal.LinkedJournal_v.DiskView./DiskView/entries (%Poly%bundle!journal.LinkedJournal_v.DiskView.
          (Poly%bundle!journal.LinkedJournal_v.DiskView. self~2@)
       )))
      ) k~12$
   )))
   :pattern ((vstd!set.impl&%0.contains.? TYPE%bundle!disk.GenericDisk_v.Address. TYPE%bundle!disk.GenericDisk_v.Address.
     (vstd!map.impl&%0.dom.? TYPE%bundle!disk.GenericDisk_v.Address. TYPE%bundle!disk.GenericDisk_v.Address.
      TYPE%bundle!journal.LinkedJournal_v.JournalRecord. TYPE%bundle!journal.LinkedJournal_v.JournalRecord.
      (Poly%vstd!map.Map<bundle!disk.GenericDisk_v.Address./bundle!journal.LinkedJournal_v.JournalRecord.>.
       (bundle!journal.LinkedJournal_v.DiskView./DiskView/entries (%Poly%bundle!journal.LinkedJournal_v.DiskView.

```

**Comment by utaal** [comment]
I don't think that commit exists in the repo (maybe it got pruned?), can you maybe push it to a branch/tag?

**Comment by jonhnet** [comment]
I was a dingus and forgot to push. Please try again.

**Comment by Chris-Hawblitzel** [comment]
@jonhnet Can you try with https://github.com/verus-lang/verus/commit/4454b0cdc9b86f84021cc4a72fa6a1a0ed603443 (which is already in the `main` branch)?

**Comment by jonhnet** [comment]
Yep, fixed in current main (post `4454b0c'). Thanks!
